pkg/addonmanager/controllers/agentdeploy/hookcontroller.go

```diff
@@ -0,0 +1,301 @@
+		}
+		finalizer := removeFinalizer(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+		managedClusterAddonCopy.SetFinalizers(finalizer)
+		_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	// has hook manifestWork
+	if !hasFinalizer(managedClusterAddonCopy.GetFinalizers(), constants.PreDeleteHookFinalizer) {
+		managedClusterAddonCopy.Finalizers = append(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+		_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	if managedClusterAddon.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// apply hookWork when addon is deleting
+	hookWork.OwnerReferences = []metav1.OwnerReference{*owner}
+	hookWork, err = c.applyWork(ctx, hookWork)
+	if err != nil {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    "HookManifestCompleted",
+			Status:  metav1.ConditionFalse,
+			Reason:  "ManifestWorkApplyFailed",
+			Message: fmt.Sprintf("failed to apply manifestwork %v: %v", hookWork.Name, err),
+		})
+		c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+		Type:    "HookManifestCompleted",
+		Status:  metav1.ConditionFalse,
+		Reason:  "HookManifestIsNotCompleted",
+		Message: fmt.Sprintf("hook manifestWork %v is not completed.", hookWork.Name),
+	})
+
+	completed := hookWorkIsCompleted(hookWork)
+	if completed {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    "HookManifestCompleted",
+			Status:  metav1.ConditionTrue,
+			Reason:  "HookManifestIsCompleted",
+			Message: fmt.Sprintf("hook manifestWork %v is completed.", hookWork.Name),
+		})
+	}
+
+	if !equality.Semantic.DeepEqual(managedClusterAddonCopy.Status, managedClusterAddon.Status) {
+		c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+	}
+
+	if completed {
+		finalizer := removeFinalizer(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+		managedClusterAddonCopy.SetFinalizers(finalizer)
+		_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	return nil
+}
+
+func (c *addonHookDeployController) applyWork(ctx context.Context, required *workapiv1.ManifestWork) (*workapiv1.ManifestWork, error) {
+	existingWork, err := c.workLister.ManifestWorks(required.Namespace).Get(required.Name)
```

Comment:
I think we should have a common apply func for both controllers.

Reply:
done